### DISCUSSION TITLE
Is there a way to programmatically save a jupyter notebook before running nbconvert?

### DISCUSSION THREAD

**Discussion by jamsden** [microsoft/vscode-jupyter#13535] [Questions and Answers]
I'm looking for a way to save a jupyter notebook that is running in vscode before invoking nbconvert to generate an HTML snapshot. I was exploring something  like:
```
import subprocess

def save_notebook():
    subprocess.run(["code", "--command", "workbench.action.files.save"])

def run_nbconvert():
    subprocess.run(["jupyter", "nbconvert", "--to", "html", "ELM_Program_Review.ipynb"])

# Save the notebook
save_notebook()

# Run nbconvert
run_nbconvert()
```
But that doesn't work - code has no --command parameter. 

Is there some other way to programmatically save the current notebook before executing nbconvert?

**Comment by ccalvocm** [comment]
Greetings @jamsden 

You can use the jupyter module to programmatically save the current notebook before executing nbconvert. Here's an example of how you could modify your code to use the jupyter module:

```
import subprocess
from jupyter_client.manager import KernelManager

def save_notebook():
    # Get the kernel ID for the current notebook
    km = KernelManager().client()
    kernel_id = km.kernel_info()['id']

    # Send a "save" message to the kernel
    km.save_connection_file()
    km.execute(kernel_id, "import os; os.kill(os.getpid(), 1)", silent=True)

def run_nbconvert():
    subprocess.run(["jupyter", "nbconvert", "--to", "html", "ELM_Program_Review.ipynb"])

# Save the notebook
save_notebook()

# Run nbconvert
run_nbconvert()
```

This code uses the jupyter_client.manager.KernelManager class to get the kernel ID for the current notebook, and then sends a "save" message to the kernel using the execute() method. This should save the current notebook before executing nbconvert.

Note that the os.kill(os.getpid(), 1) line is used to interrupt the kernel after the "save" message has been sent. This is necessary because the kernel will not respond to any further messages until the "save" operation has completed. The os.kill() call sends a SIGUSR1 signal to the current process, which interrupts the kernel and allows the nbconvert command to be executed.

**Reply by DonJayamanne** [reply]
@jamsden  thank you for filling this issue.
Unfortunately there is currently no way for the   kernel to communicate back with vscode.

If you are trying to change the behaviour of the existing export function in vscode then I think it makes sense for vscode to first  save the notebook before baking the export function, please do file an issue for that if that works for your.

**Reply by jamsden** [reply]
jupyter_client.manager.KernelManager does not seem to work when run in vscode.
